File: ruby/easy/degree_of_an_array.rb
# @param {Integer[]} nums
# @return {Integer}
def find_shortest_sub_array(nums)
    max_freq, shortest_distance = 0, nums.length
    freq_map, indexes_map = Hash.new, Hash.new
    nums.each_with_index do |num, i|
        freq = freq_map[num] ? freq_map[num]+1 : 1
        freq_map[num] = freq
        max_freq = [max_freq, freq].max
        if indexes_map[num]
            arr = indexes_map[num]
            arr[0] = [arr[0], i].min
            arr[1] = [arr[1], i].max
        else
            indexes_map[num] = [i, i]
        end
    end
    nums.each do |num|
        if freq_map[num] == max_freq
            arr = indexes_map[num]
            shortest_distance = [shortest_distance, (arr[1] - arr[0] + 1)].min
        end
    end
    
    return shortest_distance
end

File: javascript/easy/degreeOfAnArray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var findShortestSubArray = function(nums) {
    const freqMap = new Map(), indexesMap = new Map();
    let shortest = nums.length, maxFreq = 0;
    for(let i = 0; i < nums.length; i++) {
        let freq = (freqMap.has(nums[i]) ? (freqMap.get(nums[i]) + 1) : 1);
        freqMap.set(nums[i], freq);
        maxFreq = Math.max(maxFreq, freq);
        
        if(indexesMap.has(nums[i])) {
            let arr = indexesMap.get(nums[i]);
            arr.min = Math.min(arr.min, i);
            arr.max = Math.max(arr.max, i);
        } else {
            indexesMap.set(nums[i], {min: i, max: i});
        }
    }
    
    for(let i = 0; i < nums.length; i++) {
        if(freqMap.get(nums[i]) === maxFreq) {
            let arr = indexesMap.get(nums[i]);
            shortest = Math.min(shortest, (arr.max - arr.min + 1));
        }
    }
    
    return shortest;
};
